phi/torch/cuda/src/phi_torch_cuda.cpp

```diff
@@ -0,0 +1,337 @@
+        // Create sparse matrix A in CSR format
+        CHECK_CUSPARSE( cusparseCreateCsr(&matA, dim_i, dim_j, A_nnz,
+                                          dA_csrOffsets.data_ptr<int64_t>(), dA_columns.data_ptr<int64_t>(),
+                                          dA_values.data_ptr<double>(),
+                                          CUSPARSE_INDEX_64I, CUSPARSE_INDEX_64I,
+                                          CUSPARSE_INDEX_BASE_ZERO, CUDA_R_64F) )
+        // Create dense vector X
+        CHECK_CUSPARSE( cusparseCreateDnVec(&vecX, dim_j, dB.data_ptr<double>(), CUDA_R_64F) )
+        // Create dense vector y
+        torch::Tensor dC = torch::empty({dim_i}, dB.options());
+        CHECK_CUSPARSE( cusparseCreateDnVec(&vecY, dim_i, dC.data_ptr<double>(), CUDA_R_64F) )
+        // allocate an external buffer if needed
+        CHECK_CUSPARSE( cusparseSpMV_bufferSize(
+                                     handle, CUSPARSE_OPERATION_NON_TRANSPOSE,
+                                     &alpha, matA, vecX, &beta, vecY, CUDA_R_64F,
+                                     CUSPARSE_CSRMV_ALG2, &bufferSize) )
+        CHECK_CUDA( cudaMalloc(&dBuffer, bufferSize) )
+
+        // execute SpMV
+        CHECK_CUSPARSE( cusparseSpMV(handle, CUSPARSE_OPERATION_NON_TRANSPOSE,
+                                     &alpha, matA, vecX, &beta, vecY, CUDA_R_64F,
+                                     CUSPARSE_CSRMV_ALG2, dBuffer) )
+
+        // destroy matrix/vector descriptors
+        CHECK_CUSPARSE( cusparseDestroySpMat(matA) )
+        CHECK_CUSPARSE( cusparseDestroyDnVec(vecX) )
+        CHECK_CUSPARSE( cusparseDestroyDnVec(vecY) )
+        CHECK_CUSPARSE( cusparseDestroy(handle) )
+        dB = dB.view({dim_j, 1});
+        dC = dC.view({dim_i, 1});
+        return dC;
+    }
+
+    /*
+        Sparse versus Dense Matrix Multiplication.
+        This function is externally called from Python.
+    */
+    torch::Tensor cusparse_SpMM(
+                            const at::Tensor& dA_csrOffsets,
+                            const at::Tensor& dA_columns,
+                            const at::Tensor& dA_values,
+                            at::Tensor& dB,
+                            const int64_t dim_i,
+                            const int64_t dim_j)
+    {
+        if(dB.dtype() == torch::kFloat32) {
+            return floatSpMM(dA_csrOffsets, dA_columns, dA_values, dB, dim_i, dim_j);
+        }
+        else if(dB.dtype() == torch::kFloat64) {
+            return doubleSpMM(dA_csrOffsets, dA_columns, dA_values, dB, dim_i, dim_j);
+        }
+        else {
+            fprintf(stderr, "error: %s: cusparse_SpMM type undexpected at line %d\n", __FILE__,       \
+                    __LINE__);                                                     \
+            exit(1);
+        }
+    }
+
+    /*
+        Sparse versus dense matrix multiplication using CUSPARSE.
+        This function is called internally from conjugate_gradient()
```

Comment:
How does this function differ from `cusparse_SpMM` Why the prefix `__`?
Clarify in the comment.